Add Monero-Bitcoin Atomic Swap blog post (#85)
comit-network/comit.network: blog/2020-10-06-monero-bitcoin.md

---
title: "Monero-Bitcoin Atomic Swap"
author: lucas
author_url: https://github.com/luckysori
author_image_url: https://avatars3.githubusercontent.com/u/9418575
tags: [monero,bitcoin,atomic,swap,adaptor,scriptless]
---

Over half a year ago we proved that [Grin-Bitcoin atomic swaps were possible](https://github.com/comit-network/grin-btc-poc) and now we're back to show you how it can be done for the [Monero-Bitcoin pair](https://github.com/comit-network/xmr-btc-swap).
In our effort to connect all the blockchains without adding yet another, we have turned our attentions to Monero.
Similar to Grin, Monero is a public ledger in which the source, destination and amount of a transaction remain hidden to observers.

> The team has been interested in Monero for a long time, but until recently we thought that atomic swaps involving it were unfeasible for a few reasons:
> 
> 1. Timelocks are not supported, so one cannot construct spend conditions contingent on time.
> 2. It lacks a scripting language altogether, meaning that the usual mechanism of hashlocks is unavailable.
> 3. Monero's use of edwards25519 elliptic curve for its key pair generation, which limits our ability to use [adaptor signatures](https://github.com/LLFourn/one-time-VES/blob/master/main.pdf) with blockchains such as Bitcoin and Ethereum, which use a different curve.
> 
> Thanks to the publication of [Joël Gugger's Bitcoin-Monero Cross-chain Atomic Swap paper](https://eprint.iacr.org/2020/1126.pdf) we discovered ways to overcome the afforementioned challenges.

Reviewer: I think you are downplaying our capabilities a bit :D 
I'd say we knew that it is possible (Pedro published work on this) but always expected Monero to have a timelocks. 
If it's just me who reads this as we were too stupid, then leave it, otherwise let's rephrase it please :) 

My proposal (feel free to adapt): 
```suggestion
The team has been interested in Monero for a long time, but until recently we assumed that for atomic swaps both blockchains need to support timelocks. It was only through a presentation by Joël Gugger (h4sh3d) at [36C3](https://www.youtube.com/watch?v=G-v6hDnzpds&ab_channel=MoneroCommunityWorkgroup) that we learned that for atomic swaps timelocks are only required on one side of the swaps. 

Nevertheless, we knew that atomic swaps between Bitcoin and Monero are challenging because: 

1. Monero lacks a scripting language altogether, meaning that the usual mechanism of hashlocks is unavailable.
2. Monero's use of edwards25519 elliptic curve for its key pair generation, which limits our ability to use [adaptor signatures](https://github.com/LLFourn/one-time-VES/blob/master/main.pdf) with blockchains such as Bitcoin and Ethereum, which use a different curve.

Details on Joël Gugger's Bitcoin-Monero swap protocol can be found in his recently published [paper](https://eprint.iacr.org/2020/1126.pdf).

```

Reverting said change left us with a largely similar protocol with some key differences:

1. Simple 2-of-2 multisig scripts for all Bitcoin transactions.
2. Transaction-level timelocks.
3. An alternative approach to the DLEQ proof required to ensure that the correct keys are revealed when redeeming and refunding bitcoin.

With (1) and (2) we wanted to make the Bitcoin scripts as simple as possible, to uphold the privacy properties of Monero.
A 2-of-2 multisig is commonplace on Bitcoin, so it does not necessarily indicate that an atomic swap is taking place.
The third point will be discussed after the protocol has been properly discussed in the next section.

### Notation and terminology

- When referring to secret keys we use lowercase e.g. `x`; when referring to public keys we use uppercase e.g. `X`.
- We refer to Alice as the party that holds Monero and wants Bitcoin in exchange; we refer to Bob as the party that holds Bitcoin and wants Monero in exchange.

### Assumptions

We assume that Alice and Bob have exchanged a set of parameters before performing the protocol, namely:

- the amounts being exchanged; and

> - the values of the timelocks used for two of the Bitcoin transactions, `t_1` and `t_2`, where `t_2 > t_1`.

Reviewer: ```suggestion
- the values of the timelocks used for two of the Bitcoin transactions, `t_1` and `t_2`.
```

This was a left over from the old protocol. If timelocks are relative this is not needed.

So Alice does just that.
And Bob does something analogous with his secret spend key `s_b`, as that one should be leaked to Alice if he ever refunds his bitcoin, to allow her to refund her monero too.
He generates a Bitcoin public key `S_b^bitcoin` to go along with his previously sent `S_b^monero` and constructs a zero-knowledge proof showing that they both correspond to the same secret key `s_b`.

It may be obvious but each party must verify the proof provided by the other!

#### Bob constructs Bitcoin lock transaction

In order to allow Alice to sign Bitcoin transactions based on the shared output on `A` and `B`, Bob constructs the lock transaction with the help of his wallet (e.g. using bitcoind's wallet API [fundrawtransaction](https://bitcoincore.org/en/doc/0.19.0/rpc/rawtransactions/fundrawtransaction/)) and shares it with Alice, unsigned.

#### Signing worse-case scenario Bitcoin transactions

> In an ideal world, one wouldn't need to worry about refund and punish transactions, but in that case one wouldn't need atomic swaps in the first place.

Reviewer: 😅

#### Going on-chain

With Alice's cancel transaction signature and refund transaction encrypted signature, Bob can now safely lock up his bitcoin using the lock transaction he constructed before.
This is because he will be able to refund if Alice stops cooperating.

To do so, he asks his Bitcoin wallet to sign the lock transaction and broadcasts it to the network.
Alice will eventually see that the transaction has been included in a block, and may wait some extra time to see her desired number of confirmation on that transaction.

Alice can now lock up her monero, because she is safe knowing that she will be able to refund if Bob refunds, and that she will be able to just take the bitcoin if Bob never acts again.
Alice uses her Monero wallet to transfer the agreed upon monero to the output `(S_a + S_b, V_a + V_b)`.
Bob will be able to see that this has taken place, because he knows `S_a`, `S_b`, `V_a` and `V_b`, and also `v_a` and `v_b`, granting him the ability to see transactions paying to that output.
If he verifies that the amount of monero is correct and the transaction has enough confirmations, he can proceed with the protocol.

#### One last signature

> The current situation allows both parties to abort without any further cooperation, but Alice does not yet have a way to take Bob's bitcoin.

Reviewer 2: Except after `t2`, right?

Author: Yep, do you think it's worth clarifying?

Reviewer 2: I balked at it, perhaps 

```suggestion
The current situation allows both parties to abort without any further cooperation, but Alice does not yet have a way to take Bob's bitcoin (before the punish timelock).
```
Feel free to not use this though.

Given our unfamiliarity with ring signatures, a construct central to the protocol, and the apparent lack of an implementation in Rust, our programming language of choice, our ex-colleague Lloyd Fournier advised us to instead [turn to the textbook for answers](https://www.win.tue.nl/~berry/CryptographicProtocols/LectureNotes.pdf)

Lloyd proposed an equivalent proof based on the well-defined composition of sigma protocols.
In short, for every bit of the secret key, one can prove that said bit is either equal to 0 for the Bitcoin public key **and** the Monero public key, *or* equal to 1 for the Bitcoin public key **and** the Monero public key.

For more details on this, check out [the source code of our experimental implementation of this cross-curve DLEQ proof protocol](https://github.com/comit-network/cross-curve-dleq).

## Just show me the code

Having explained how we think this can work, it is time to show you what we've done.

> Over the last couple of weeks, part of the team has been working on a proof-of-concept pure-Rust implementation of this protocol.

Reviewer: The actual link to our PoC is missing :D 
```suggestion
Over the last couple of weeks, part of the team has been working on a [proof-of-concept pure-Rust implementation](https://github.com/comit-network/xmr-btc-swap/) of this protocol.
```

> Using cryptographic libraries such as [curve25519-dalek](https://github.com/dalek-cryptography/curve25519-dalek) and [secp256kFUN!](https://github.com/LLFourn/secp256kfun/), together with the aforementioned [cross-curve DLEQ proof library]((https://github.com/comit-network/cross-curve-dleq)), we have built a library which can be used to atomically swap Monero and Bitcoin.

Reviewer 2: Two sets of parenthesis here `((https://github.com/ ...`

We must emphasise that a considerable part of the cryptography used has not been audited or thoroughly reviewed yet, so we recommend anyone curious to use this library to only use it with as much real money as they are willing to risk losing.

## What's next

With the publication of this post marking the end of this project, we asked ourselves what we could do next.
Here are some possibilities:

- Building an experimental peer-to-peer application so that people can use it to trade test coins or small amounts of real coins.
- Extend our recent [implementation](https://github.com/comit-network/thor) of [generalised Bitcoin-compatible channels](https://eprint.iacr.org/2020/476.pdf) with support for atomic swaps between Monero and off-chain Bitcoin.
- Move on to a different cryptocurrency pair altogether, and possibly apply some of the learnings garnered during this project.

In any case, we will likely get to all of those at some point in the future.

> For the time being, thank you for reading!

Reviewer: Add: 

* feedback are welcome as tickets or on the maillinglist

We also assume that parties have a means of communicating with each other during the protocol.

### Long story short

<img alt="BTC/XMR Atomic Swap Protocol" src={useBaseUrl('/blog/assets/images/2020-10/BTC_XMR atomic swap protocol.svg')} />;

In the best-case scenario the protocol looks like this:

1. Alice and Bob exchange a set of addresses, keys, zero-knowledge proofs and signatures.
2. Bob publishes `Tx_fund` and locks up his bitcoin in a 2-of-2 multisig output owned by Alice and Bob (`A+B`).
Given the information exchanged in step 1, Bob can refund his bitcoin if he waits until time `t_1`.
If Bob doesn't refund after time `t_1` using `Tx_cancel` and `Tx_refund`, Alice can punish Bob for being inactive by spending first `Tx_cancel` and after `t_2` spending the output using `Tx_punish`.
3. Alice sees that Bob has locked up the bitcoin, so she publishes `Tx_lock` on the monero blockchain and hence locks up her monero in an output which can only be spent with a secret key owned by Alice (`s_a`) *and* a secret key owned by Bob (`s_b`).
This means that neither of them can actually spend this output unless they learn the secret key of the other party.

> 4. Bob sees that Alice has locked up the monero, so he now sends Alice a missing key bit of information which will allow Alice to redeem the Bitcoin (`Tx_redeem`).

Reviewer 3: What is this `missing key bit of information`?

With (1) and (2) we wanted to make the Bitcoin scripts as simple as possible, to uphold the privacy properties of Monero.
A 2-of-2 multisig is commonplace on Bitcoin, so it does not necessarily indicate that an atomic swap is taking place.
The third point will be discussed after the protocol has been properly discussed in the next section.

### Notation and terminology

- When referring to secret keys we use lowercase e.g. `x`; when referring to public keys we use uppercase e.g. `X`.
- We refer to Alice as the party that holds monero and wants bitcoin in exchange; we refer to Bob as the party that holds bitcoin and wants monero in exchange.

### Assumptions

We assume that Alice and Bob have exchanged a set of parameters before performing the protocol, namely:

- the amounts being exchanged; and
- the values of the absolute timelocks used for two of the Bitcoin transactions, `t_1` and `t_2`, where `t_2 > t_1`.

We also assume that parties have a means of communicating with each other during the protocol.

### Long story short
In the best-case scenario the protocol looks like this:

> <img alt="BTC/XMR Atomic Swap Protocol" src={useBaseUrl('/blog/assets/images/2020-10/BTC_XMR atomic swap protocol.svg')} />

Reviewer 3: This is not rendered on the preview.

Reviewer: I know, still fighting with this :S

Reviewer: I think I fixed it :)
At least it works locally: 

<img width="742" alt="image" src="https://user-images.githubusercontent.com/224613/95169577-d44f8180-07fe-11eb-8303-490204afd6d8.png">

So Alice does just that.
And Bob does something analogous with his secret spend key `s_b`, as that one should be leaked to Alice if he ever refunds his bitcoin, to allow her to refund her monero too.
He generates a Bitcoin public key `S_b^bitcoin` to go along with his previously sent `S_b^monero` and constructs a zero-knowledge proof showing that they both correspond to the same secret key `s_b`.

It may be obvious but each party must verify the proof provided by the other!

#### Bob constructs Bitcoin lock transaction

In order to allow Alice to sign Bitcoin transactions based on the shared output on `A` and `B`, Bob constructs `Tx_lock` with the help of his wallet (e.g. using bitcoind's wallet API [fundrawtransaction](https://bitcoincore.org/en/doc/0.19.0/rpc/rawtransactions/fundrawtransaction/)) and shares it with Alice, unsigned.

> #### Signing worse-case scenario Bitcoin transactions
> In an ideal world, one wouldn't need to worry about refund and punish transactions, but in that case one wouldn't need atomic swaps in the first place.

Reviewer 3: ```suggestion
#### Signing worse-case scenario Bitcoin transactions

In an ideal world, one wouldn't need to worry about refund and punish transactions, but in that case one wouldn't need atomic swaps in the first place.
```
just to match other titles
